controllers/controllers.go

```diff
@@ -41,19 +41,61 @@ type instanceReconciler struct {
 	recorder record.EventRecorder
 }
 
-// configureInstance adds the specified instance to the cluster. if hostname is not empty, the instance's hostname will be
-// changed to the passed in value. If annotations is not nil, the node will have the specified annotations applied to
-// it.
-func (r *instanceReconciler) configureInstance(instance *instances.InstanceInfo, annotations map[string]string) error {
+// ensureInstanceIsUpToDate ensures that the given instance is configured as a node and upgraded to the specifications
+// defined by the current version of WMCO. If annotations is not nil, the node will have the specified annotations
+// applied to it.
+func (r *instanceReconciler) ensureInstanceIsUpToDate(instance *instances.InstanceInfo, annotationsToApply map[string]string) error {
+	if instance == nil {
+		return errors.New("instance cannot be nil")
+	}
+
+	// Instance is up to date, do nothing
+	if instance.Node != nil && instanceUpToDate(instance.Node.Annotations) {
+		return nil
+	}
+
 	nc, err := nodeconfig.NewNodeConfig(r.k8sclientset, r.clusterServiceCIDR, r.vxlanPort, instance, r.signer,
-		annotations)
+		annotationsToApply)
 	if err != nil {
 		return errors.Wrap(err, "failed to create new nodeconfig")
 	}
-	if err := nc.Configure(); err != nil {
-		return errors.Wrap(err, "failed to configure Windows instance")
+
+	// Check if the instance was configured by a previous version of WMCO and must be deconfigured before being
+	// configured again.
+	if instance.Node != nil && upgradeRequired(instance.Node.Annotations) {
+		if err := nc.Deconfigure(); err != nil {
+			return err
+		}
 	}
-	return nil
+
+	return nc.Configure()
+}
+
+// instanceUpToDate returns true if the instance was configured by the current WMCO version
+func instanceUpToDate(nodeAnnotations map[string]string) bool {
+	versionAnnotation, _ := nodeAnnotations[nodeconfig.VersionAnnotation]
+	return versionAnnotation == version.Get()
+}
+
+// upgradeRequired returns true if the instance needs to go through the upgrade process.
+func upgradeRequired(nodeAnnotations map[string]string) bool {
+	annotation, present := nodeAnnotations[nodeconfig.VersionAnnotation]
+
+	// Version annotation not being present means that the node has been created but not fully configured.
+	// The upgrade process is not required, the node should just be configured normally.
+	if !present {
+		return false
+	}
+
+	// Version annotation equal to the current version means that the node has been fully configured by this version
+	// of WMCO. No further action is required.
```

Comment:
Is `No further action is required` accurate? Returning `false` here just implies that the instance will not be `deconfigured` but the instance will go through  configuration.

Reply:
Ill rephrase